### ISSUE TITLE
cmp_nan only compares against NaN constant

### ISSUE THREAD

**Issue by killercup** [rust-lang/rust-clippy#1205]
IIUC, the `cmp_nan` lint just checks whether an expression is a constant whose last path segement is `NAN`.

That means that

``` rust
0 == std::f32::NAN; //~ERROR doomed comparison with NAN
```

works as expected but more complicated (but still rather trivial) stuff like

``` rust
let z = 42_f64;

let am_i_nan = std::f64::NAN;
z == am_i_nan; //~ERROR doomed comparison with NAN

let am_i_nan = std::f64::NEG_INFINITY / std::f64::INFINITY;
z == am_i_nan; //~ERROR doomed comparison with NAN
```

does not produce any of the expected errors.

Since the lint is only supposed to work with those standard number representations that have a `NaN` value (i.e. f32, f64, f128), the number of cases that evaluate to NaN is hopefully not that large and can probably be hard coded.

**Comment by killercup** [comment]
Oh, or am I misunderstanding the lint? Is this so people write `…::is_nan()` to be more explicit?

**Comment by mcarton** [comment]
> Is this so people write `…::is_nan()` to be more explicit?

It's not to be more explicit, it's just that `foo == NAN` does not work, NaN equals nothing, not even itself.

**Comment by killercup** [comment]
> NaN equals nothing, not even itself.

Doh! I totally knew that… I'll grab a non-alcoholic drink next.

**Comment by Lokathor** [comment]
So... should we close this?

**Comment by flip1995** [comment]
This lint could still be improved, by also evaluating `const`s: [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d2f5a3ad3d884642e30d221fcc6f5439)

```rust
const MY_NAN: f64 = std::f64::NAN;

fn main() {
    let x = 3.0f64;
    
    if x == MY_NAN { // This currently doesn't get linted
        println!("YaY");
    } else if x == std::f64::NAN {
        println!("NaY")
    }
}
```
